### DISCUSSION TITLE
Consideration of common probes and VO-specific probes

### DISCUSSION THREAD

**Discussion by rdimaio** [rucio/rucio#7084] [General]
Continuation of the thread found here: https://github.com/rucio/rucio/issues/6717

**Comment by dchristidis** [comment]
> I don't see any other official Rucio components (from [this list](https://rucio.github.io/documentation/component_leads/)) that are VO-specific,

The Lifetime Model is ATLAS-specific. It has the potential to become VO-agnostic, if anyone else is interested, but it’s not. Similarly for the Consistency Checks (two competing implementations, one for ATLAS, one for CMS). And in the RSE Decommissioner (which is not yet decided whether it may become an independent component, it is envisioned that the VO-specific profiles be kept together with the common profile.

> should we also have a single repository for all VO-specific policy package plugins?

That was always my thought and I assumed that it wasn’t done so for packaging reasons. But beyond the collaboration, there is another argument in favour: support and debugging. When a community is having trouble, it takes additional effort to remember whether they have a policy package, and even more so to locate it.

> do they all implement monitoring in some way?

No, probes do all sorts of things. Monitoring and other type of checks is one. Others are housekeeping (e.g. some RSE usage counters) and synchronisation with external services (e.g. accounts, RSEs). It can be the appropriate location for anything that has to run periodically, but doesn’t fit in any daemon.

> It might make sense to abstract out the Rucio monitoring logic into the probe so that it is all in one place.

How do you mean? I cannot imagine combining all the monitoring probes in some way being beneficial, but likely I have misunderstood your suggesting.

> Regarding Reaper, does it depend on a probe that implements monitoring? If not, it could be safe to implement that probe's logic directly into Reaper

Part of what I wrote turns out to be outdated. The Reaper used to depend on the `obsolete` RSE usage counter. It appears that that has been removed.

By default, a Reaper is suboptimal (must be run in greedy mode). The way to not unnecessarily delete data is to collection the RSE occupancy from the storage directly. However, there’s no VO-agnostic method to do this.

But it comes back to what I wrote before: nobody managed to propose an interesting method to have periodically-running procedures inside a daemon itself. And then you have to worry about whether such a procedure may take too long, or what to do when having multiple daemon instances.

The probes turned out to be the solution with the least amount of friction.
